### PR TITLE
Stream `.rrd` files

### PR DESCRIPTION
**Pull request by jprochazk**
### What

Closes https://github.com/rerun-io/rerun/issues/2262

Instead of waiting for the `.rrd` file to finish downloading, `stream_rrd_from_http` now uses `ehttp::streaming` and parses messages as data is streamed in.

As part of this change, compression now uses the LZ4 block format instead of the LZ4 frame format. **This is a breaking change to the RRD format**.

| example                | `main`     | `jan/stream-rrd` | change  |
| ---------------------- | ---------- | ---------------- | ------- |
| signed_distance_fields | 21.328 MB  | 21.420 MB        | +0.43%  |
| deep_sdf               | 21.329 MB  | 21.329 MB        | 0.00%   |
| dicom                  | 67.657 MB  | 67.657 MB        | 0.00%   |
| raw_mesh               | 109.070 KB | 110.960 KB       | +1.73%  |
| clock                  | 54.901 KB  | 80.439 KB        | +46.52% |
| dicom_mri              | 67.656 MB  | 66.434 MB        | -1.81%  |
| api_demo               | 320.549 KB | 388.021 KB       | +21.05% |
| structure_from_motion  | 252.978 MB | 253.735 MB       | +0.30%  |
| car                    | 402.542 KB | 391.056 KB       | -2.85%  |
| rgbd                   | 168.420 MB | 167.382 MB       | -0.62%  |
| colmap                 | 253.054 MB | 253.054 MB       | 0.00%   |
| plots                  | 149.360 KB | 211.155 KB       | +41.37% |
| nyud                   | 168.467 MB | 168.467 MB       | 0.00%   |
| text_logging           | 2.732 KB   | 3.091 KB         | +13.14% |

### Checklist
* [ ] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [ ] I've included a screenshot or gif (if applicable)

PR Build Summary: https://build.rerun.io/pr/2412

Docs preview: https://rerun.io/preview/0c82335/docs
Examples preview: https://rerun.io/preview/0c82335/examples

### CODE REVIEW

**Comment by emilk** [comment]
That `nyud` change looks very suspicious - are you sure you didn't accidentally test it without the new `jpeg_quality` feature?

**Comment by emilk** [comment]
I think its time to add a docstring to `re_log_encoding/src/lib.rs` explaining how messages are encoded on the wire

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+        let mut decoder = StreamDecoder::new();
+        for chunk in data.chunks(1) {
+            decoder.push_chunk(chunk.to_vec());
+        }
+
+        for _ in 0..16 {
+            assert_message_ok!(decoder.try_read());
+        }
+    }
```
Comment (emilk) [inline crates/re_log_encoding/src/decoder/stream.rs:280]: you should also test that you can:

* call `decoder.try_read()` without first pushing any chunks (and get `None`)
* push a few chunks, get a few things out of `decoder.try_read()`, then `None`, and then push the rest of the chunks and read the rest of the messages (i.e. actually stream!)
* push chunks of various sizes (not just the edge-case of size `1`, which never tests the `Cursor` part of `Chunk`

**Review by emilk — COMMENTED** [review]
I've done a partial review, and it looks really good so far!

But I think `ChunkBuffer` has some nasty bugs in it, and I suspect it can be simplified quite a bit.

File: crates/re_log_encoding/src/stream_rrd_from_http.rs [hunk 2]
```diff
@@ -42,27 +44,84 @@ pub type HttpMessageCallback = dyn Fn(HttpMessage) + Send + Sync;
...
+                        ));
+                        ControlFlow::Break(())
+                    }
+                }
+                ehttp::streaming::Part::Chunk(chunk) => {
+                    if chunk.is_empty() {
+                        on_msg(HttpMessage::Success);
+                        return ControlFlow::Break(());
+                    }
+
```
Comment (emilk) [inline crates/re_log_encoding/src/stream_rrd_from_http.rs:73]:
```suggestion

                    re_tracing::profile_scope!("decoding_rrd_stream");
```

File: crates/re_log_encoding/src/stream_rrd_from_http.rs [hunk 3]
```diff
@@ -42,27 +44,84 @@ pub type HttpMessageCallback = dyn Fn(HttpMessage) + Send + Sync;
...
         Err(err) => {
             on_msg(HttpMessage::Failure(
-                format!("Failed to fetch .rrd file from {url}: {err}").into(),
+                format!("Failed to decode .rrd: {err}").into(),
             ));
         }
-    });
-}
+    }
+} */
```
Comment (emilk) [inline crates/re_log_encoding/src/stream_rrd_from_http.rs:124]: rm dead code

**Review by emilk — CHANGES_REQUESTED** [review]

File: crates/re_log_encoding/src/decoder.rs [hunk 4]
```diff
@@ -60,71 +64,50 @@ pub fn decode_bytes(bytes: &[u8]) -> Result<Vec<LogMsg>, DecodeError> {
...
-            Compression::Off => Self::Uncompressed(read),
-            Compression::LZ4 => Self::Lz4(lz4_flex::frame::FrameDecoder::new(read)),
-        }
+    let FileHeader {
+        magic,
+        version,
+        options,
+    } = FileHeader::decode(&mut read)?;
+
+    if &magic == b"RRF0" {
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder.rs]: Bumped to `RRF2` and put the old versions in a constant array which the magic is checked against here

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    /// The stream state
+    state: State,
+}
+
+#[derive(Clone, Copy)]
+enum State {
+    StreamHeader,
+    MessageHeader,
+    Message(MessageHeader),
+}
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder/stream.rs]: Added an ascii flow and a comment on each variant explaining what it's for

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+type Chunk = Cursor<Vec<u8>>;
+
+struct ChunkBuffer {
+    /// Any incoming chunks are queued until they are emptied
+    queue: VecDeque<Chunk>,
+
+    /// When `try_read` is called and we don't have enough bytes yet,
+    /// we store whatever we do have in this buffer
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder/stream.rs:142]: The data is copied into the buffer, and the chunk retains it, but it won't be read ever again, because the `Chunk` cursor position is updated. The buffer exists so that `try_read` can return a contiguous slice.

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+    /// Attempt to read `n` bytes out of the queued chunks.
+    ///
+    /// Returns `Ok(None)` if there is not enough data to return a slice of `n` bytes.
+    fn try_read(&mut self, n: usize) -> Result<Option<&[u8]>, DecodeError> {
+        // resize the buffer if the target has changed
+        if self.buffer.len() != n {
+            self.buffer.resize(n, 0);
+            self.cursor = 0;
+        }
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder/stream.rs:170]: I think asserting against it would require adding a field tracking the `n` so we can detect when the sequence of events you describe happens. It's only used internally here, so would documenting it be enough?

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    }
+
+    fn push(&mut self, chunk: Vec<u8>) {
+        self.queue.push_back(Chunk::new(chunk));
+    }
+
+    /// Attempt to read `n` bytes out of the queued chunks.
+    ///
+    /// Returns `Ok(None)` if there is not enough data to return a slice of `n` bytes.
+    fn try_read(&mut self, n: usize) -> Result<Option<&[u8]>, DecodeError> {
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder/stream.rs:165]: True, it's reading from `Cursor<Vec<u8>>`, and if that's empty it will just repeatedly return `Ok(0)` (at which point the chunk will be removed), so I'll replace this with an `expect` and make `try_read` infallible

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+fn is_chunk_empty(chunk: &Chunk) -> bool {
+    chunk.position() >= chunk.get_ref().len() as u64
+}
+
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder/stream.rs:208]: You were right, `chunk_buffer_read_same_n` now tests for this

**Comment by jprochazk** [comment]
I'll also re-test the file sizes now that `main` is merged into here

File: crates/re_log_encoding/src/decoder/stream.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_log_encoding/src/decoder/stream.rs]: Add that as a docstring - future readers of the code will ask the same question!

File: crates/re_log_encoding/src/decoder/stream.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_log_encoding/src/decoder/stream.rs]:
```diff
-self.cursor = 0;
+assert_eq!(self.cursor, 0);
```

should do it, no?

**Comment by jprochazk** [comment]
| example                | `main`     | `jan/stream-rrd` | change  |
| ---------------------- | ---------- | ---------------- | ------- |
| signed_distance_fields | 21.328 MB  | 21.420 MB        | +0.43%  |
| deep_sdf               | 21.329 MB  | 21.329 MB        | 0.00%   |
| dicom                  | 67.657 MB  | 67.657 MB        | 0.00%   |
| raw_mesh               | 109.070 KB | 110.960 KB       | +1.73%  |
| clock                  | 54.901 KB  | 80.439 KB        | +46.52% |
| dicom_mri              | 67.656 MB  | 66.434 MB        | -1.81%  |
| api_demo               | 320.549 KB | 388.021 KB       | +21.05% |
| structure_from_motion  | 252.978 MB | 253.735 MB       | +0.30%  |
| car                    | 402.542 KB | 391.056 KB       | -2.85%  |
| rgbd                   | 168.420 MB | 167.382 MB       | -0.62%  |
| colmap                 | 253.054 MB | 253.054 MB       | 0.00%   |
| plots                  | 149.360 KB | 211.155 KB       | +41.37% |
| nyud                   | 168.467 MB | 168.467 MB       | 0.00%   |
| text_logging           | 2.732 KB   | 3.091 KB         | +13.14% |

Looks a lot better now. There seem to be a few outliers, but they're measured in `KB`.

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+struct ChunkBuffer {
+    /// Any incoming chunks are queued until they are emptied
+    queue: VecDeque<Chunk>,
+
+    /// When `try_read` is called and we don't have enough bytes yet,
+    /// we store whatever we do have in this buffer
+    buffer: Vec<u8>,
+
+    /// The cursor points to the end of the used range in `buffer`
+    cursor: usize,
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder/stream.rs]: Renamed this to `buffer_fill` and updated the comment

File: crates/re_log_encoding/src/decoder/stream.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/re_log_encoding/src/decoder/stream.rs]: True, added the assert

File: crates/re_log_encoding/Cargo.toml [hunk 11]
```diff
@@ -37,7 +37,9 @@ re_smart_channel.workspace = true
 re_tracing.workspace = true
 
 # External:
-ehttp = "0.2"
+ehttp = { git = "https://github.com/jprochazk/ehttp.git", features = [
+  "streaming",
+] }
```
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_log_encoding/Cargo.toml:42]: https://github.com/emilk/ehttp/pull/28 has been merged, so we can update this now

**Comment by emilk** [comment]
It would be nice with a debug-level log when we reach the end of the HTTP stream. It will make testing the streaming behavior easier

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+    fn new() -> Self {
+        Self {
+            queue: VecDeque::with_capacity(16),
+            buffer: Vec::with_capacity(1024),
+            buffer_fill: 0,
+        }
+    }
+
+    fn push(&mut self, chunk: Vec<u8>) {
+        self.queue.push_back(Chunk::new(chunk));
```
Comment (emilk) [inline crates/re_log_encoding/src/decoder/stream.rs:161]: maybe early-out here for empty chunks

File: crates/re_log_encoding/src/decoder/stream.rs [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+
+        // keep pushing 3 chunks of 16 bytes at a time, and attempting to read messages
+        // until there are no more chunks
+        let mut chunks = data.chunks(16).peekable();
+        while chunks.peek().is_some() {
+            for _ in 0..3 {
+                if chunks.peek().is_none() {
+                    break;
+                }
+                decoder.push_chunk(chunks.next().unwrap().to_vec());
```
Comment (emilk) [inline crates/re_log_encoding/src/decoder/stream.rs:375]: no need for unwrap

```suggestion
                if let Some(chunk) = chunks.next() {
                    decoder.push_chunk(chunk.to_vec());
                } else {
                    break;
                }
```

**Review by emilk — APPROVED** [review]
Code LGTM (except for wrong error message, see above), but there is something going wrong when testing this (again, see above). I added the `do-no-merge` label until the testing works.

Also: please update the PR description with the latest (correct) table, and mention the breaking change to the RRD stream format.

**Comment by jprochazk** [comment]
> I'm testing this with:

I just tested this and it seems to work for me. The `build_demo_app` script does not build the SDK, but uses it to render the data files. It seems like that could be the problem here, could you run the SDK build, `rm -rf web_demo` and then try again?

File: crates/re_log_encoding/Cargo.toml [hunk 14]
```diff
@@ -37,7 +37,7 @@ re_smart_channel.workspace = true
 re_tracing.workspace = true
 
 # External:
-ehttp = "0.2"
+ehttp = { git = "https://github.com/emilk/ehttp.git", features = ["streaming"] }
```
Comment (emilk) [inline crates/re_log_encoding/Cargo.toml:40]:
```suggestion
ehttp = { git = "https://github.com/emilk/ehttp.git", features = ["streaming"], rev = "884e25a0835ab4cbea53180cb61660100866f28a" }
```

otherwise the build can break when we check out this `rerun` commit months from now (e.g. during a `git bisect`)

**Review by emilk — APPROVED** [review]

**Comment by emilk** [comment]
Seems to be working as advertised 🥳
